src/services/concretes/lottery.service.ts

```diff
@@ -0,0 +1,102 @@
+import { DataAccessLayer } from "@blockr/blockr-data-access";
+import { logger } from "@blockr/blockr-logger";
+import { Block, State } from "@blockr/blockr-models";
+import { inject, injectable } from "inversify";
+import * as seedRandom from "seedrandom";
+import { LotteryException } from "../../exceptions/lotteryException.exception";
+import { QueueStore } from "../../stores/queue.stores";
+
+@injectable()
+export class LotteryService {
+  private readonly dataAccessLayer: DataAccessLayer;
+  private readonly queueStore: QueueStore;
+  private ticketCount: number;
+
+  constructor(@inject(DataAccessLayer) dataAccessLayer: DataAccessLayer,
+              @inject(QueueStore) queueStore: QueueStore) {
+    this.dataAccessLayer = dataAccessLayer;
+    this.queueStore = queueStore;
+    this.ticketCount = 0;
+  }
+
+  public async drawWinningBlock(parentBlockHash: string,
+                                walletStates: State[]): Promise<Block> {
+      return new Promise(async (resolve, reject) => {
+        const pendingProposedBlocks: Set<Block> = this.queueStore.pendingProposedBlockQueue;
+        const stakeMap = await this.convertStatesToStakeMap(walletStates);
+
+        try {
+          const candidatesMap = await this.calculateCandidates(stakeMap, pendingProposedBlocks);
+
+          resolve(await this.chooseWinningBlock(parentBlockHash, candidatesMap, pendingProposedBlocks));
+        } catch (error) {
+          logger.error(error);
+          
+          reject(error);
+        }
+      });
+  }
+
+  private async chooseWinningBlock(parentBlockHash: string, candidatesMap: Map<string, number>,
+                                   pendingProposedBlocks: Set<Block>): Promise<Block> {
+    return new Promise((resolve, reject) => {
+        // TODO: Perhaps the random stake choosing should be overhauled.
+        const randomNumber = seedRandom(parentBlockHash)() * this.ticketCount;
+        let bottomMargin = 0;
+        let upperMargin = 0;
+        let candidateIndex = 0;
+
+        while (upperMargin < randomNumber) {
+          const publicKey = Array.from(candidatesMap.keys())[candidateIndex];
+          // The stake can never be undefined in this part of the code as candidates are only inserted into the map
+          // whenever they have an actual stake, thus the exclamation mark.
+          const stake = candidatesMap.get(publicKey) as number;
+          upperMargin += stake;
+
+          if ((bottomMargin <= randomNumber) && (upperMargin >= randomNumber)) {
+            resolve(Array.from(pendingProposedBlocks).find((block) => block.blockHeader.validator === publicKey));
+          }
+          bottomMargin += stake;
+          candidateIndex += 1;
+        }
+        reject(new LotteryException("The lottery does not yield any winner for the given proposedblocks."));
```

Comment:
proposedblocks are 2 separate words? Otherwise you shouldn't call the variable names ProposedBlocks (with capital B)